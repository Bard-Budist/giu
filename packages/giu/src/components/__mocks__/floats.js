/* eslint-env jest */
const Floats = jest.genMockFromModule('../floats');

Floats.Floats = mockComponent('Floats');

module.exports = Floats;
