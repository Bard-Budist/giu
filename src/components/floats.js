import React                from 'react';
import { createStore }      from 'redux';
import {
  addLast, removeAt,
  set as timmSet,
  merge, mergeIn, addDefaults,
  omit,
}                           from 'timm';
import {
  bindAll,
  cancelEvent,
  windowHeightWithoutScrollbar, windowWidthWithoutScrollbar,
}                           from '../gral/helpers';
import { isVisible }        from '../gral/visibility';
import { MISC }             from '../gral/constants';
import {
  boxWithShadow,
}                           from '../gral/styles';

const PROP_KEYS_TO_REMOVE_FROM_FLOAT_DIV = [
  'position', 'align', 'zIndex', 'limitSize', 'getAnchorNode',
];

// ==========================================
// Store, reducer
// ==========================================
let store = null;
function initStore() { store = createStore(reducer); }

const INITIAL_STATE = {
  cntReposition: 0,
  floats: [],
};
function reducer(state0 = INITIAL_STATE, action) {
  let state = state0;
  let idx;
  let id;
  switch (action.type) {
    case 'FLOAT_ADD':
      state = timmSet(state, 'floats', addLast(state.floats, action.pars));
      state = timmSet(state, 'cntReposition', state.cntReposition + 1);
      break;
    case 'FLOAT_DELETE':
      id = action.id;
      idx = state.floats.findIndex(o => o.id === id);
      if (idx >= 0) {
        state = timmSet(state, 'floats', removeAt(state.floats, idx));
      }
      break;
    case 'FLOAT_UPDATE':
      id = action.id;
      idx = state.floats.findIndex(o => o.id === id);
      if (idx >= 0) {
        state = mergeIn(state, ['floats', idx], action.pars);
      }
      break;
    case 'FLOAT_REPOSITION':
      state = timmSet(state, 'cntReposition', state.cntReposition + 1);
      break;
    default:
      break;
  }
  return state;
}

// ==========================================
// Action creators
// ==========================================
let cntId = 0;
const DEFAULT_FLOAT_PARS = {
  zIndex: MISC.zMainFloatDelta,
  limitSize: false,
};
const actions = {
  floatAdd: initialPars => {
    const id = `float_${cntId++}`;
    const pars = addDefaults(initialPars, DEFAULT_FLOAT_PARS, { id });
    return { type: 'FLOAT_ADD', pars };
  },
  floatDelete: id => ({ type: 'FLOAT_DELETE', id }),
  floatUpdate: (id, pars) => ({ type: 'FLOAT_UPDATE', id, pars }),
  floatReposition: () => ({ type: 'FLOAT_REPOSITION' }),
};

// Imperative dispatching
const floatAdd = pars => {
  const action = actions.floatAdd(pars);
  store.dispatch(action);
  return action.pars.id;
};
const floatDelete = (id) => store.dispatch(actions.floatDelete(id));
const floatUpdate = (id, pars) => store.dispatch(actions.floatUpdate(id, pars));
const floatReposition = () => store && store.dispatch(actions.floatReposition());

// ==========================================
// Inits
// ==========================================

// Reposition all floats upon window scroll or resize
// (This does *not* cover div scrolling, of course -- for that,
// you need to explicitly attach the `floatReposition` dispatcher
// as a listener to the corresponding `scroll` event)
try {
  window.addEventListener('scroll', floatReposition);
  window.addEventListener('resize', floatReposition);
} catch (err) { /* ignore */ }

// ==========================================
// Position and visibility calculation
// ==========================================
function isAnchorVisible({ getAnchorNode }) {
  const anchorNode = getAnchorNode();
  if (!anchorNode) return null;
  return isVisible(anchorNode);
}

// ==========================================
// Floats component
// ==========================================
let fFloatsMounted = false;
const isFloatsMounted = () => fFloatsMounted;

class Floats extends React.PureComponent {

  constructor(props) {
    super(props);
    bindAll(this, [
      'forceUpdate',
      'renderFloat',
    ]);
    if (!store) initStore();
    this.storeUnsubscribe = store.subscribe(this.forceUpdate);
  }

  componentWillMount() {
    fFloatsMounted = true;
    this.refFloats = [];
  }

  componentWillUnmount() {
    fFloatsMounted = false;
    if (this.storeUnsubscribe) this.storeUnsubscribe();
  }

  componentDidUpdate() {
    const { curState, prevState } = this;
    if (curState.cntReposition !== prevState.cntReposition) {
      this.repositionFloats();
    }
  }

  // ==========================================
  render() {
    this.prevState = this.curState;
    this.curState = store.getState();
    this.floats = this.curState.floats;
    return (
      <div
        className="giu-floats"
        style={style.outer}
      >
        {this.floats.map(this.renderFloat)}
      </div>
    );
  }

  renderFloat(props, idx) {
    if (!isAnchorVisible(props)) return null;
    const { id, zIndex } = props;
    return (
      <div key={id}
        className="giu-float"
        onMouseDown={cancelEvent}
        style={style.wrapper(zIndex)}
      >
        <div ref={c => { this.refFloats[idx] = c; }}
          {...omit(props, PROP_KEYS_TO_REMOVE_FROM_FLOAT_DIV)}
          style={style.floatInitial(props)}
        />
      </div>
    );
  }

  // ==========================================
  repositionFloats() {
    // Make it asynchronous, so that we are not in the middle
    // of React's update cycle (which may cause some refs to be
    // dettached and floats not to be positioned properly).
    // We use a promise here, since it triggers a microtask
    // (faster than a task)
    Promise.resolve().then(() => {
    // setTimeout(() => {
      const floats = this.floats;
      for (let idx = 0; idx < floats.length; idx++) {
        this.repositionFloat(floats[idx], idx);
      }
    });
  }

  repositionFloat(float, idx) {
    const ref = this.refFloats[idx];
    if (!ref) return;

    // Hide and move to top-left for measuring
    let { position, align } = float;
    const { limitSize } = float;
    ref.style.opacity = '0.01';
    ref.style.left = '0px';
    ref.style.right = null;
    ref.style.top = '0px';
    ref.style.bottom = null;
    if (limitSize) {
      ref.style.overflowX = 'visible';
      ref.style.overflowY = 'visible';
      ref.style.maxWidth = null;
      ref.style.maxHeight = null;
    }
    const wFloat = ref.offsetWidth;
    const hFloat = ref.offsetHeight;

    // Preparations
    const anchorNode = float.getAnchorNode();
    const bcr = anchorNode && anchorNode.getBoundingClientRect();
    if (!bcr) return;
    const styleAttrs = {};
    const hWin = windowHeightWithoutScrollbar();
    const wWin = windowWidthWithoutScrollbar();
    const breathe = MISC.windowBorderBreathe;

    // Position vertically
    if (!position) {
      const freeBelow = hWin - bcr.bottom;
      if (freeBelow >= hFloat) {
        position = 'below';
      } else {
        position = freeBelow > bcr.top ? 'below' : 'above';
      }
    }
    let maxHeight;
    if (position === 'below') {
      styleAttrs.top = `${bcr.bottom}px`;
      styleAttrs.bottom = null;
      maxHeight = Math.max(0, hWin - bcr.bottom - breathe);
    } else {
      styleAttrs.top = null;
      styleAttrs.bottom = `${windowHeightWithoutScrollbar() - bcr.top}px`;
      maxHeight = Math.max(0, bcr.top - breathe);
    }
    if (limitSize) styleAttrs.overflowY = 'auto';

    // Position horizontally
    if (!align) {
      const freeRight = wWin - bcr.left;
      if (freeRight >= wFloat) {
        align = 'left';
      } else {
        align = freeRight > bcr.right ? 'left' : 'right';
      }
    }
    let maxWidth;
    if (align === 'left') {
      styleAttrs.left = `${bcr.left}px`;
      styleAttrs.right = null;
      maxWidth = Math.max(0, wWin - bcr.left - breathe);
    } else {
      styleAttrs.left = null;
      styleAttrs.right = `${windowWidthWithoutScrollbar() - bcr.right}px`;
      maxWidth = Math.max(0, bcr.right - breathe);
    }
    if (limitSize) styleAttrs.overflowX = 'auto';

    // Limit size. This must be at the end, so that the changes are also
    // applied at the end
    if (limitSize) {
      styleAttrs.maxHeight = hFloat > maxHeight ? `${maxHeight}px` : null;
      styleAttrs.maxWidth = wFloat > maxWidth ? `${maxWidth}px` : null;
    }

    // Apply style
    Object.keys(styleAttrs).forEach(attr => {
      ref.style[attr] = styleAttrs[attr];
    });
    ref.style.opacity = 1;
  }
}

// ==========================================
// Styles
// ==========================================
const style = {
  outer: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: 0,
    height: 0,
  },
  wrapper: zIndex => ({
    position: 'fixed',
    top: 0,
    left: 0,
    width: 0,
    height: 0,
    zIndex,
  }),
  floatInitial: ({ style: baseStyle, noStyleShadow }) => {
    let out = {
      position: 'fixed',
      top: 0,
      left: 0,
      opacity: 0.5,
    };
    if (!noStyleShadow) out = boxWithShadow(out);
    if (baseStyle) out = merge(out, baseStyle);
    return out;
  },
};

// ==========================================
// Warnings
// ==========================================
let fCheckedFloats = false;
const floatsWarning = name => `<${name}> requires a <Floats> component to be \
included in your application. It will not work properly otherwise. Please add it \
as close as possible to the application root; no props are needed.`;

function warnFloats(componentName) {
  if (fCheckedFloats) return;
  fCheckedFloats = true;
  /* eslint-disable no-console */
  if (!isFloatsMounted()) console.warn(floatsWarning(componentName));
  /* eslint-enable no-console */
}


// ==========================================
// Public API
// ==========================================
export {
  Floats,
  floatAdd, floatDelete, floatUpdate, floatReposition,
  warnFloats,
};
