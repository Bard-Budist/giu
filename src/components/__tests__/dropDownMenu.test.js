// @flow

/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import DropDownMenu from '../dropDownMenu';
import { LIST_SEPARATOR } from '../../';
import { floatAdd, floatDelete } from '../floats';

const lang = 'es';
const ITEMS = [
  { label: () => (lang === 'es' ? 'Manzanas' : 'Apples'), value: 'a' },
  { label: '2', value: 2 },
  LIST_SEPARATOR,
  { label: () => (lang === 'es' ? 'Bayas' : 'Berries'), value: 'b' },
  { label: 'true', value: true },
  { label: 'C', value: 'c' },
];

// https://github.com/facebook/react/issues/7386#issuecomment-238091398
jest.mock('react-dom');
jest.mock('../floats');

describe('DropDownMenu', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('01 renders (closed)', () => {
    const tree = renderer.create(
      <DropDownMenu items={ITEMS}>Menu title</DropDownMenu>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('02 toggles highlight on/off when clicked upon', () => {
    const component = renderer.create(
      <DropDownMenu items={ITEMS}>Menu title</DropDownMenu>
    );
    let tree;
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // Open
    const focusEl = $(tree, 'input');
    focusEl.props.onFocus();
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // Close
    focusEl.props.onBlur();
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('03 shows/hides float when clicked upon', () => {
    floatAdd.mockReturnValue(425);
    const component = renderer.create(
      <DropDownMenu items={ITEMS}>Menu title</DropDownMenu>
    );
    const tree = component.toJSON();
    const focusEl = $(tree, 'input');

    // Open
    focusEl.props.onFocus();
    expect(floatAdd.mock.calls.length).toEqual(1);
    const float = renderer.create(floatAdd.mock.calls[0][0].children).toJSON();
    expect(float).toMatchSnapshot();

    // Close
    focusEl.props.onBlur();
    expect(floatDelete.mock.calls.length).toEqual(1);
    expect(floatDelete.mock.calls[0][0]).toEqual(425);
  });
});
