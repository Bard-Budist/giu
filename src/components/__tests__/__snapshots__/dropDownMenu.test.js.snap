exports[`DropDownMenu 01 renders (closed) 1`] = `
<span
  className="giu-select-custom"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "display": "inline-block",
      "position": "relative",
    }
  }>
  <input
    onBlur={[Function]}
    onCopy={[Function]}
    onCut={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    style={
      Object {
        "cursor": "default",
        "height": 0,
        "left": 8,
        "opacity": 0,
        "padding": 0,
        "pointerEvents": "none",
        "position": "fixed",
        "top": 8,
        "width": 0,
        "zIndex": -80,
      }
    }
    tabIndex={undefined} />
  <div
    className="giu-drop-down-menu-title"
    onMouseDown={[Function]}
    style={
      Object {
        "backgroundColor": undefined,
        "color": undefined,
        "cursor": "pointer",
        "display": "inline-block",
      }
    }>
    Menu title
  </div>
</span>
------------HTML PREVIEW---------------
<span
  class="giu-select-custom"
  style="display:inline-block;position:relative;"
>
  <input style="cursor:default;height:0;left:8px;opacity:0;padding:0;pointer-events:none;position:fixed;top:8px;width:0;z-index:-80;">
  <div
    class="giu-drop-down-menu-title"
    style="cursor:pointer;display:inline-block;"
  >
    Menu title
  </div>
</span>
`;

exports[`DropDownMenu 02 toggles highlight on/off when clicked upon 1`] = `
<span
  className="giu-select-custom"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "display": "inline-block",
      "position": "relative",
    }
  }>
  <input
    onBlur={[Function]}
    onCopy={[Function]}
    onCut={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    style={
      Object {
        "cursor": "default",
        "height": 0,
        "left": 8,
        "opacity": 0,
        "padding": 0,
        "pointerEvents": "none",
        "position": "fixed",
        "top": 8,
        "width": 0,
        "zIndex": -80,
      }
    }
    tabIndex={undefined} />
  <div
    className="giu-drop-down-menu-title"
    onMouseDown={[Function]}
    style={
      Object {
        "backgroundColor": undefined,
        "color": undefined,
        "cursor": "pointer",
        "display": "inline-block",
      }
    }>
    Menu title
  </div>
</span>
------------HTML PREVIEW---------------
<span
  class="giu-select-custom"
  style="display:inline-block;position:relative;"
>
  <input style="cursor:default;height:0;left:8px;opacity:0;padding:0;pointer-events:none;position:fixed;top:8px;width:0;z-index:-80;">
  <div
    class="giu-drop-down-menu-title"
    style="cursor:pointer;display:inline-block;"
  >
    Menu title
  </div>
</span>
`;

exports[`DropDownMenu 02 toggles highlight on/off when clicked upon 2`] = `
<span
  className="giu-select-custom"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "display": "inline-block",
      "position": "relative",
    }
  }>
  <input
    onBlur={[Function]}
    onCopy={[Function]}
    onCut={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    style={
      Object {
        "cursor": "default",
        "height": 0,
        "left": 8,
        "opacity": 0,
        "padding": 0,
        "pointerEvents": "none",
        "position": "fixed",
        "top": 8,
        "width": 0,
        "zIndex": -80,
      }
    }
    tabIndex={undefined} />
  <div
    className="giu-drop-down-menu-title"
    onMouseDown={[Function]}
    style={
      Object {
        "backgroundColor": "#6600cc",
        "color": "white",
        "cursor": "pointer",
        "display": "inline-block",
      }
    }>
    Menu title
  </div>
</span>
------------HTML PREVIEW---------------
<span
  class="giu-select-custom"
  style="display:inline-block;position:relative;"
>
  <input style="cursor:default;height:0;left:8px;opacity:0;padding:0;pointer-events:none;position:fixed;top:8px;width:0;z-index:-80;">
  <div
    class="giu-drop-down-menu-title"
    style="background-color:#6600cc;color:white;cursor:pointer;display:inline-block;"
  >
    Menu title
  </div>
</span>
`;

exports[`DropDownMenu 02 toggles highlight on/off when clicked upon 3`] = `
<span
  className="giu-select-custom"
  onClick={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "display": "inline-block",
      "position": "relative",
    }
  }>
  <input
    onBlur={[Function]}
    onCopy={[Function]}
    onCut={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    style={
      Object {
        "cursor": "default",
        "height": 0,
        "left": 8,
        "opacity": 0,
        "padding": 0,
        "pointerEvents": "none",
        "position": "fixed",
        "top": 8,
        "width": 0,
        "zIndex": -80,
      }
    }
    tabIndex={undefined} />
  <div
    className="giu-drop-down-menu-title"
    onMouseDown={[Function]}
    style={
      Object {
        "backgroundColor": undefined,
        "color": undefined,
        "cursor": "pointer",
        "display": "inline-block",
      }
    }>
    Menu title
  </div>
</span>
------------HTML PREVIEW---------------
<span
  class="giu-select-custom"
  style="display:inline-block;position:relative;"
>
  <input style="cursor:default;height:0;left:8px;opacity:0;padding:0;pointer-events:none;position:fixed;top:8px;width:0;z-index:-80;">
  <div
    class="giu-drop-down-menu-title"
    style="cursor:pointer;display:inline-block;"
  >
    Menu title
  </div>
</span>
`;

exports[`DropDownMenu 03 shows/hides float when clicked upon 1`] = `
<div
  className="giu-list-picker"
  onWheel={[Function]}
  style={
    Object {
      "backgroundColor": "white",
      "border": "1px solid #999999",
      "color": "inherit",
      "fontFamily": "inherit",
      "fontSize": "inherit",
      "fontWeight": "inherit",
      "maxHeight": "inherit",
      "overflowY": "auto",
      "paddingBottom": 3,
      "paddingTop": 3,
    }
  }>
  <div
    id="\"a\""
    onClick={undefined}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "alignItems": "center",
        "border": "1px solid transparent",
        "cursor": "default",
        "display": "flex",
        "flexDirection": "row",
        "padding": "3px NaNpx 3px 10px",
        "whiteSpace": "nowrap",
      }
    }>
    Manzanas
  </div>
  <div
    id="2"
    onClick={undefined}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "alignItems": "center",
        "border": "1px solid transparent",
        "cursor": "default",
        "display": "flex",
        "flexDirection": "row",
        "padding": "3px NaNpx 3px 10px",
        "whiteSpace": "nowrap",
      }
    }>
    2
  </div>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "default",
        "height": 7,
        "paddingBottom": 3,
        "paddingTop": 3,
      }
    }>
    <div
      style={
        Object {
          "borderTop": "1px solid #999999",
          "height": 1,
        }
      } />
  </div>
  <div
    id="\"b\""
    onClick={undefined}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "alignItems": "center",
        "border": "1px solid transparent",
        "cursor": "default",
        "display": "flex",
        "flexDirection": "row",
        "padding": "3px NaNpx 3px 10px",
        "whiteSpace": "nowrap",
      }
    }>
    Bayas
  </div>
  <div
    id="true"
    onClick={undefined}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "alignItems": "center",
        "border": "1px solid transparent",
        "cursor": "default",
        "display": "flex",
        "flexDirection": "row",
        "padding": "3px NaNpx 3px 10px",
        "whiteSpace": "nowrap",
      }
    }>
    true
  </div>
  <div
    id="\"c\""
    onClick={undefined}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "alignItems": "center",
        "border": "1px solid transparent",
        "cursor": "default",
        "display": "flex",
        "flexDirection": "row",
        "padding": "3px NaNpx 3px 10px",
        "whiteSpace": "nowrap",
      }
    }>
    C
  </div>
</div>
------------HTML PREVIEW---------------
<div
  class="giu-list-picker"
  style="background-color:white;border:1px solid #999999;color:inherit;font-family:inherit;font-size:inherit;font-weight:inherit;max-height:inherit;overflow-y:auto;padding-bottom:3px;padding-top:3px;"
>
  <div
    id="\"a\""
    style="align-items:center;border:1px solid transparent;cursor:default;display:flex;flex-direction:row;padding:3px NaNpx 3px 10px;white-space:nowrap;"
  >
    Manzanas
  </div>
  <div
    id="2"
    style="align-items:center;border:1px solid transparent;cursor:default;display:flex;flex-direction:row;padding:3px NaNpx 3px 10px;white-space:nowrap;"
  >
    2
  </div>
  <div style="cursor:default;height:7px;padding-bottom:3px;padding-top:3px;">
    <div style="border-top:1px solid #999999;height:1px;"></div>
  </div>
  <div
    id="\"b\""
    style="align-items:center;border:1px solid transparent;cursor:default;display:flex;flex-direction:row;padding:3px NaNpx 3px 10px;white-space:nowrap;"
  >
    Bayas
  </div>
  <div
    id="true"
    style="align-items:center;border:1px solid transparent;cursor:default;display:flex;flex-direction:row;padding:3px NaNpx 3px 10px;white-space:nowrap;"
  >
    true
  </div>
  <div
    id="\"c\""
    style="align-items:center;border:1px solid transparent;cursor:default;display:flex;flex-direction:row;padding:3px NaNpx 3px 10px;white-space:nowrap;"
  >
    C
  </div>
</div>
`;
